src/objective-c/GRPCClient/GRPCCall.m

```diff
@@ -794,112 +793,114 @@ - (void)invokeCall {
             }
             userInfo[kGRPCTrailersKey] = strongSelf.responseTrailers;
-            // TODO(jcanizales): The C gRPC library doesn't guarantee that the headers block will be
-            // called before this one, so an error might end up with trailers but no headers. We
-            // shouldn't call finishWithError until ater both blocks are called. It is also when
-            // this is done that we can provide a merged view of response headers and trailers in a
-            // thread-safe way.
-            if (strongSelf.responseHeaders) {
-              userInfo[kGRPCHeadersKey] = strongSelf.responseHeaders;
-            }
+            // Since gRPC core does not guarantee the headers block being called before this block,
+            // responseHeaders might be nil.
+            userInfo[kGRPCHeadersKey] = strongSelf.responseHeaders;
             error = [NSError errorWithDomain:error.domain code:error.code userInfo:userInfo];
           }
-          [strongSelf maybeFinishWithError:error];
+          [strongSelf finishWithError:error];
+          strongSelf->_requestWriter.state = GRXWriterStateFinished;
         }
       }];
-  // Now that the RPC has been initiated, request writes can start.
-  @synchronized(_requestWriter) {
-    [_requestWriter startWithWriteable:self];
-  }
 }
 
 #pragma mark GRXWriter implementation
 
+// Lock acquired inside startWithWriteable:
 - (void)startCallWithWriteable:(id<GRXWriteable>)writeable {
-  _responseWriteable =
-      [[GRXConcurrentWriteable alloc] initWithWriteable:writeable dispatchQueue:_responseQueue];
-
-  GRPCPooledChannel *channel =
-      [[GRPCChannelPool sharedInstance] channelWithHost:_host callOptions:_callOptions];
-  GRPCWrappedCall *wrappedCall = [channel wrappedCallWithPath:_path
-                                              completionQueue:[GRPCCompletionQueue completionQueue]
-                                                  callOptions:_callOptions];
-
-  if (wrappedCall == nil) {
-    [self maybeFinishWithError:[NSError errorWithDomain:kGRPCErrorDomain
-                                                   code:GRPCErrorCodeUnavailable
-                                               userInfo:@{
-                                                 NSLocalizedDescriptionKey :
-                                                     @"Failed to create call or channel."
-                                               }]];
-    return;
-  }
-
   @synchronized(self) {
-    _wrappedCall = wrappedCall;
-  }
+    if (_state == GRXWriterStateFinished) {
+      return;
+    }
+
+    _responseWriteable =
+        [[GRXConcurrentWriteable alloc] initWithWriteable:writeable dispatchQueue:_responseQueue];
+
+    GRPCPooledChannel *channel =
+        [[GRPCChannelPool sharedInstance] channelWithHost:_host callOptions:_callOptions];
+    _wrappedCall = [channel wrappedCallWithPath:_path
+                                completionQueue:[GRPCCompletionQueue completionQueue]
```

Comment:
Should move ` [channel wrappedCallWithPath]` out of ` @synchronized(self){}`?

```
  @synchronized(self) {
    if (_state == GRXWriterStateFinished) {
      return;
    }
  }

  GRPCPooledChannel *channel =
      [[GRPCChannelPool sharedInstance] channelWithHost:_host callOptions:_callOptions];
  GRPCWrappedCall *wrappedCall = [channel wrappedCallWithPath:_path
                                completionQueue:[GRPCCompletionQueue completionQueue]
                                    callOptions:_callOptions];
  @synchronized(self) {
    if (_state == GRXWriterStateFinished) {
      return;
    }
    _responseWriteable =
        [[GRXConcurrentWriteable alloc] initWithWriteable:writeable dispatchQueue:_responseQueue];
    _wrappedCall = wrappedCall;
    if (_wrappedCall == nil) {
      [self finishWithError:[NSError errorWithDomain:kGRPCErrorDomain
                                                code:GRPCErrorCodeUnavailable
                                            userInfo:@{
                                              NSLocalizedDescriptionKey :
                                                  @"Failed to create call or channel."
                                            }]];
      return;
    }
    [self sendHeaders];
    [self invokeCall];
    ...
  }

Reply:
Resolved during offline discussion.

Reply:
Chatted offline. It should stay in the block (no deadlock issue; needs to guarantee state consistency of `GRPCCall`)